### ISSUE TITLE
Gundam: Senjou No Kizuna Portable Adhoc network bugs

### ISSUE THREAD

**Issue by shanzenos** [hrydgard/ppsspp#14172]
This issue has been occurring for years now, although at the moment I am fully updated to PPSSPP 1.11.2
When entering the multiplayer lobby, matches work for only 2 players- and works fine. If even one more player enters the lobby while 2 are already present, they will "join" and be able to view the lobby, but will not show up in the player list. If the match is started all players will get a connection error message and the room will close it's self. Multiplayer for just 2 players works perfectly fine however. The game supports a maximum of 8 players over adhoc.

**Comment by ANR2ME** [comment]
@shanzenos could you test with multiple instances of PPSSPP on the same PC with more than 2 players? To see if it works or not

**Comment by ANR2ME** [comment]
@shanzenos could you guide me how to do multiplayer? how to host & join?
I can't read japanese text LOL
![image](https://user-images.githubusercontent.com/7974720/108524167-56010100-7301-11eb-9813-b4657889cd3e.png)

**Comment by shanzenos** [comment]
I've heard up to 8 players used to work in older versions of the emulator but I was never given a version number. I'm thankful though for you doing some debugging for the problem, hopefully the attention eventually gets the bug fixed I know quite a large number of people that would be thrilled to hear about it being fixed.

**Comment by sum2012** [comment]
@ANR2ME 
For the blue screen Try
in Core\PSPloaders.cpp
Force Memory::g_MemorySize = Memory::RAM_DOUBLE_SIZE;

**Comment by ANR2ME** [comment]
> 
> 
> @ANR2ME
> For the blue screen Try
> in Core\PSPloaders.cpp
> Force Memory::g_MemorySize = Memory::RAM_DOUBLE_SIZE;

do i need to change the source code like that or can i do it from compat.ini ?

**Comment by sum2012** [comment]
The change is wrong,Just a test whether is memory issue 
> > @ANR2ME
> > For the blue screen Try
> > in Core\PSPloaders.cpp
> > Force Memory::g_MemorySize = Memory::RAM_DOUBLE_SIZE;
> 
> do i need to change the source code like that or can i do it from compat.ini ?

**Comment by MojoJojoDojo** [comment]
Not related to the issue but if you try to save a replay that shows Android VS PC you might see funny animations on the Android side lol.
PC to PC will show the replays correctly though.

**Comment by MojoJojoDojo** [comment]
Seems like 3 players works fine now if you wait a bit for the 3rd player to show up at least locally.
I wonder if having force real clock sync on could improve it further maybe.

**Comment by PAUV0S** [comment]
but 3 players seems to work locally although needs more testing to see if it also works on vpn

**Comment by MojoJojoDojo** [comment]
For some reason it only works with a VPN.
Trying a public server will just end in misery for all players involved.

**Comment by ANR2ME** [comment]
Hmm.. when i tried to test this again using the fix i did for Super Pocket Tennis on my test build, it seems all 4 players are visible on each PPSSPP, but if i just leave it there one of the PPSSPP got frozen, which will eventually causes the other players to shows disconnection dialog
![image](https://user-images.githubusercontent.com/7974720/134105955-a59e26c6-8aee-4c22-b924-0ed4806de32c.png)
i can't even tell whether this is a regression or an improvement LOL

PS: I'm not why it became frozen, as the log didn't show anything, and showing the Disassembly window didn't seems to hit any breakpoint, nor a crash, pressing ESC to go back to PPSSPP settings didn't work either :(

**Comment by ANR2ME** [comment]
The issue that caused Failed to allocate memory seems to be due to piling up AdhocMatching Events (which came from incoming data from players), which may lead to invalid memory access when `null` address from the failing allocation being passed to the callback handler.
![image](https://user-images.githubusercontent.com/7974720/134144563-1a5c2b53-1fb5-4b26-80ec-65ff9a1265c2.png)
```
AdhocMatching - Remaining Events: 932 <-- this is way too much, normally about 1-2 pending events in the queue for each additional player
```
Each events have an optional data of 768 bytes which is allocated on PSP memory (rounded to 1024 bytes i guess), and with more players more data will be received, thus more events will be generated.

Something similar was also happened on Lord of Arcana in the past, and the fix i did for it wasn't perfect.
I think the only way to properly fix it is probably by implementing ACK and blocking current thread on `sceNetAdhocMatchingSendData` while waiting for the ACK event, this might prevent the game from flooding too much data due to faked ACK.

**Comment by MojoJojoDojo** [comment]
I think this is related to games that generate a random SSID for each session/player (noticed this on xlink).
If so then Ace Combat: Joint Assault gonna have a similar issue as well.

**Comment by ANR2ME** [comment]
Ugh.. i also found crash issue at `return hleLogSuccessVerboseX(SCENET, ERROR_NET_ADHOC_WOULD_BLOCK, "would block");` of `sceNetAdhocPdpRecv` :(
![image](https://user-images.githubusercontent.com/7974720/134223761-47e95ec5-3f5b-4958-a9fd-ab9480ee84ae.png)
I guess `hleLog` isn't thread-safe and shouldn't be called from another real thread..

**Comment by hrydgard** [comment]
Yeah, hleLog should only happen on the main CPU emulation thread I think... It looks odd to have another thread call a HLE function directly (sceNetAdhocPdpRecv)? Maybe some code should be extracted from that function into a smaller one, that doesn't call hleLog?

**Comment by ANR2ME** [comment]
Well the MatchingInput (which is listening on incoming PDP packets and generate the Events) and MatchingEvent threads was supposed to be running on PSP side, but because the code was probably ported directly from pro-online client it ended running on the emulator side.

I'm planning to change those threads into PSPThreads later, probably after 1.12 is out, as it's too risky to do it near the end of the milestone.

**Comment by MojoJojoDojo** [comment]
> Ugh.. i also found crash issue at `return hleLogSuccessVerboseX(SCENET, ERROR_NET_ADHOC_WOULD_BLOCK, "would block");` of `sceNetAdhocPdpRecv` :(
> ![image](https://user-images.githubusercontent.com/7974720/134223761-47e95ec5-3f5b-4958-a9fd-ab9480ee84ae.png)
> I guess `hleLog` isn't thread-safe and shouldn't be called from another real thread.. in this case the MatchingInput thread

When that crash happens? I didn't notice any crashes...

**Comment by ANR2ME** [comment]
It only crashed with the fix i made for Super Pocket Tennis, which seems to fix the issue on this game too

**Comment by ANR2ME** [comment]
Hmm.. it seems the crash issue on the Log was because i'm printing the syscall's argument.

If i use `DEBUG_LOG(SCENET, "%08x=sceNetAdhocPdpRecv(%i): would block", ERROR_NET_ADHOC_WOULD_BLOCK, id);`  it will also crashed, probably because `id` is part of the syscall's arg

But if i use `DEBUG_LOG(SCENET, "%08x=sceNetAdhocPdpRecv: would block", ERROR_NET_ADHOC_WOULD_BLOCK);` it won't crash.

**Comment by ANR2ME** [comment]
Btw @MojoJojoDojo if a player joined a room and then left the room, is that player's name still appeared on the other player's list?
 
Like this 4th player that are no longer in the room (which i feels odd)
![image](https://user-images.githubusercontent.com/7974720/134452775-50f866bd-f176-45a8-992c-91b3124b63a7.png)
The 4th player's name will only disappeared when the 4th player is exiting multiplayer mode.
Is this also happening on PSP?

**Comment by ANR2ME** [comment]
Ugh.. i bumped into another crash issue i'm not sure what's happening
![image](https://user-images.githubusercontent.com/7974720/134455018-f3532322-a61c-4b59-82ff-80f9172db804.png)
![image](https://user-images.githubusercontent.com/7974720/134455099-a38cadd1-ec66-4b89-ae15-89ddbf1f249e.png)
Is this happening because i allocate userMemory from a different thread? (in this case the MatchingEvent thread)

**Comment by MojoJojoDojo** [comment]
> Btw @MojoJojoDojo if a player joined a room and then left the room, is that player's name still appeared on the other player's list?
> 
> Like this 4th player that are no longer in the room (which i feels odd)
> ![image](https://user-images.githubusercontent.com/7974720/134452775-50f866bd-f176-45a8-992c-91b3124b63a7.png)
> The 4th player's name will only disappeared when the 4th player is exiting multiplayer mode.
> Is this also happening on PSP?

When I tried it on Xlink I dont think it happened.
On xlink you cant get it to work it seems like when starting the game the game will be stuck in some sort of an endless loop (is this what they call infinite timeout?).

**Comment by unknownbrackets** [comment]
Yes, PSP memory allocation is not safe to do from another thread.

In C++, everything is unsafe to do from another thread unless you've specifically done a lot of work to make it safe (and paid a performance penalty for this.)  STL containers like std::vector?  Unsafe.  BlockAllocator?  Unsafe.  CoreTiming ticks?  Unsafe.

The places we use threads all have to have the data communicated to and from separated so they're never cross-talking with other threads.

-[Unknown]

**Comment by MojoJojoDojo** [comment]
After you finish 1 round on coop at least then all the names show up (even the 3rd player) so the inital lobby is some sort of bug.

**Comment by ANR2ME** [comment]
@shanzenos please try again after this PR got merged https://github.com/hrydgard/ppsspp/pull/15331
Or you can use the artifact from that PR to test it.

**Comment by MojoJojoDojo** [comment]
Using localhost only 4 players will work now.
5 players and above still fail to show properly.

**Comment by ANR2ME** [comment]
This game is using AdhocMatching intensively (similar to Lord of Arcana), thus for each additional player there will be more events being queue at a rapid pace :( 

I guess i'll need to find a way to process those matchingEvents faster, or at least prevent them from interfering a syscall that waited for a specific event.

**Comment by MojoJojoDojo** [comment]
Actually if you are super fast at inputting , if you access the lobby in split second time with 7 instances you can get the lobby to be almost full.
You need to make the controller to work at the same time across all the instances though.
The game is not that demanding at the lobby so I think its feasible 
Intense stuff indeed lol.

**Comment by shanzenos** [comment]
> This game is using AdhocMatching intensively (similar to Lord of Arcana), thus for each additional player there will be more events being queue at a rapid pace :(
> 
> I guess i'll need to find a way to process those matchingEvents faster, or at least prevent them from interfering a syscall that waited for a specific event.
> 
> PS: i'm not sure whether my laptop can handle more than 4 instances of PPSSPP to test 8 players LOL

@anr2me 
Hey just curious, I read about your recent changes with v1.20, is there a chance the work done there may finally get this working properly?

**Comment by shanzenos** [comment]
Just a heads up, after testing this out with the new Relay server system, I was able to confirm it works just fine with full lobbies. In fact most of the games that had issues before seem to be fully working now. Great work to everyone who worked on the networking updates, very much appreciated.

**Comment by hrydgard** [comment]
Cool, thanks for testing! I'm gonna go ahead and close this.
